[Merged by Bors] - feat(data/(d)finsupp,fintype,fun_like): add fintype and (in)finite instances

alexjbest opened the pull request:
---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

vihdzp commented on src/data/finsupp/basic.lean:
```diff
@@ -337,6 +337,11 @@ begin
 end
 
+-- really we only need one infinite and the others inhabited
+instance [infinite M] [nonempty α] : infinite (α →₀ M) :=
```
Since `infinite` is `Prop`-valued, you could also add the instance for `nontrivial M` and `infinite α`. Does this version exist for `pi` types too?

alexjbest replied:
Sounds good, will add.

vihdzp replied:
Here you go:
```lean
instance [infinite α] [nontrivial M] : infinite (α →₀ M) :=
let ⟨m, hm⟩ := exists_ne (0 : M) in infinite.of_injective _ $ finsupp.single_left_injective hm
```

vihdzp commented on src/data/finsupp/basic.lean:
```diff
@@ -337,6 +337,11 @@ begin
+-- really we only need one infinite and the others inhabited
+instance [infinite M] [nonempty α] : infinite (α →₀ M) :=
+infinite.of_injective (λ i, finsupp.single (classical.arbitrary _) i)
+  (finsupp.single_injective (classical.arbitrary _))
```
Does this work?
```suggestion
infinite.of_injective _ $ finsupp.single_injective $ classical.arbitrary α
```

vihdzp commented on src/data/finsupp/basic.lean:
```diff
@@ -337,6 +337,11 @@ begin
   exact nontrivial_of_ne (single default x) 0 (mt single_eq_zero.1 hx)
 end
 
+-- really we only need one infinite and the others inhabited
```
I don't get what this remark is referring to. If you mean `α` and `M`, note that `ℕ →₀ fin 1` is not infinite, even though `ℕ` is infinite and `fin 1` is inhabited.

alexjbest replied:
This is a copy paste from the other one, but it doesn't really make sense here, it would for dfinsupp as I mean we only need one `i` such that `\pi i` is infinite, the others only need to be inhabited

alexjbest commented on src/data/finsupp/basic.lean:
```diff
@@ -337,6 +337,10 @@ begin
   exact nontrivial_of_ne (single default x) 0 (mt single_eq_zero.1 hx)
 end
 
+-- really we only need one infinite and the others inhabited
```
```suggestion
```

vihdzp commented on src/data/fintype/basic.lean:
```diff
@@ -2010,6 +2010,13 @@ infinite.of_surjective (coe : list α → multiset α) (surjective_quot_mk _)
 instance [infinite α] : infinite (option α) :=
 infinite.of_injective some (option.some_injective α)
 
+-- really we only need one infinite and the others inhabited
```
I wonder if perhaps you could prove this as a lemma, and use it to build the less general instance.

Vierkantor committed:
feat(data/{dfinsupp,finsupp,fintype}): `fintype` and `infinite` instances for `pi`, `finsupp`, `dfinsupp`

Some little instances we needed for defining the ideal norm.

Co-Authored-By: Alex J. Best <alex.j.best@gmail.com>

Vierkantor committed:
Add module doc

Vierkantor committed:
Make this instance more computable (but still noncomputable because all `finsupp` definitions are noncomputable)

Vierkantor committed:
Add the left infinite case for `finsupp`

Co-Authored-By: Violeta Hernández <vi.hdz.p@gmail.com>

Vierkantor committed:
Add the left infinite case for `dfinsupp`

Vierkantor committed:
Add the left infinite case for `pi`

Vierkantor committed:
Don't need this `haveI` anymore

Vierkantor committed:
Lint fix

alreadydone commented on src/data/dfinsupp/basic.lean:
```diff
@@ -1929,3 +1935,24 @@ add_monoid_hom.congr_fun (comp_lift_add_hom h.to_add_monoid_hom g) f
+instance dfinsupp.infinite_of_left {ι : Sort*} {π : ι → Sort*}
+  [∀ i, nontrivial (π i)] [Π i, has_zero (π i)] [infinite ι] :
+  infinite (Π₀ i, π i) :=
+by letI := classical.dec_eq ι; choose m hm using (λ i, exists_ne (0 : π i)); exact
```
We usually just use the `classical` tactic, but I guess there's no harm to be more specific :)

Vierkantor replied:
I like to be precise with my use of classical, if doesn't make everything too verbose anyway. Here specifically I think it's a good idea because `dfinsupp` is designed to be mostly computable.

alreadydone replied:
Sounds good!

alreadydone commented on src/data/dfinsupp/basic.lean:
```diff
@@ -1929,3 +1935,24 @@ add_monoid_hom.congr_fun (comp_lift_add_hom h.to_add_monoid_hom g) f
+
+section finite_infinite
+
+instance dfinsupp.fintype {ι : Sort*} {π : ι → Sort*} [decidable_eq ι] [Π i, has_zero (π i)]
```
Should we add the `finite` instances as well? `finite` is equivalent to `nonempty fintype`, and I've thought about automatically getting `nonempty` instance [here](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/.E2.9C.94.20gcd_monoid/near/303539302), though there's one more thing to consider here: `[decidable_eq ι]` wouldn't be needed for the `finite` instances.

alreadydone commented on src/data/fintype/basic.lean:
```diff
@@ -2130,6 +2130,34 @@ begin
   exact H'.false
 end
 
+instance pi.infinite_of_left {ι : Type*} {π : ι → Sort*} [∀ i, nontrivial $ π i]
```
I guess we still want to specialize to non-dependent functions because Lean sometimes has problem finding the `∀`/pi instances?

alreadydone commented on src/data/fintype/basic.lean:
```diff
@@ -2130,6 +2130,34 @@ begin
+  simp only [eq_self_iff_true, if_true] at this,
+  split_ifs at this, { assumption },
+  have := hm y,
+  contradiction
```
```suggestion
  choose m n hm using (λ i, exists_pair_ne (π i)),
  refine infinite.of_injective (λ i, m.update i (n i)) (λ x y h, not_not.1 $ λ hne, _),
  simp_rw [update_eq_iff, update_noteq hne] at h,
  exact (hm x h.1.symm).elim,
```

alreadydone commented on src/data/fintype/basic.lean:
```diff
@@ -2130,6 +2130,34 @@ begin
+(infinite.of_injective (λ (p : π i) t,
+  if h : t = i then cast (congr_arg π h.symm) p
+  else classical.arbitrary _)
+  (λ x y h, by simpa using congr_fun h i))
```
```suggestion
let ⟨m⟩ := @pi.nonempty ι π _ in infinite.of_injective _ (update_injective m i)
```

alreadydone commented on src/data/mv_polynomial/cardinal.lean:
```diff
@@ -26,7 +27,7 @@ variables {σ : Type u} {R : Type v} [comm_semiring R]
+  letI := classical.dec_eq σ,
   refine (mk_finsupp_lift_of_infinite _ R).trans _,
   rw [mk_finsupp_nat, max_assoc, lift_max, lift_aleph_0, max_comm],
 end
```
No need of `decidable_eq` now (probably after you fixed the linting errors):
```suggestion
(mk_finsupp_lift_of_infinite _ R).trans $
  by rw [mk_finsupp_nat, max_assoc, lift_max, lift_aleph_0, max_comm]
```

Vierkantor committed:
Golf

Co-authored-by: Junyan Xu <junyanxumath@gmail.com>

Vierkantor committed:
Golf

Co-authored-by: Junyan Xu <junyanxumath@gmail.com>

Vierkantor committed:
Add a `finite` instance for `dfinsupp`

This is slightly weaker but might be more convenient if we don't care about computing.

Vierkantor committed:
Add non-dependent `infinite` instances for pi types

alreadydone commented on src/data/dfinsupp/basic.lean:
```diff
@@ -1929,3 +1935,31 @@ add_monoid_hom.congr_fun (comp_lift_add_hom h.to_add_monoid_hom g) f
+/-- Weaker form of `dfinsupp.fintype`. -/
+instance dfinsupp.finite {ι : Sort*} {π : ι → Sort*} [Π i, has_zero (π i)]
+  [finite ι] [∀ i, finite (π i)] :
+  finite (Π₀ i, π i) :=
```
Thanks for adding this, but the finsupp version is missing right? And should we also add `dfinsupp.infinite_of_right` for completeness? [pi.finite](https://leanprover-community.github.io/mathlib_docs/data/finite/basic.html#pi.finite) does exist, but `function.finite` and `function.fintype` are also missing: maybe that's indication that the `pi` versions are enough and `function.infinite_of_left/right` are actually not necessary?

Vierkantor replied:
> And should we also add `dfinsupp.infinite_of_right` for completeness?

This was already in the first commit, did you mean something else?

> [pi.finite](https://leanprover-community.github.io/mathlib_docs/data/finite/basic.html#pi.finite) does exist, but `function.finite` and `function.fintype` are also missing: maybe that's indication that the `pi` versions are enough and `function.infinite_of_left/right` are actually not necessary?

I don't think it's strictly necessary to add them, especially since Lean 4 fixes the dependent instance bug that would require these duplications. So we could just hope that nobody triggers the bug in this context before we migrate to Lean 4. But now the instances are already here, I don't see an objection to keeping them because that would be much less work than adding them again when the dependent instance bug shows up.

alreadydone replied:
Sorry 😅 It seems what's missing is:
```
lemma dfinsupp.infinite_of_exists_right {ι : Type*} {π : ι → Type*} 
  (i : ι) [infinite $ π i] [Π i, has_zero (π i)] :
  infinite (Π₀ i : ι, π i) :=
```

alreadydone commented on src/data/fintype/basic.lean:
```diff
@@ -2130,6 +2130,36 @@ begin
+instance pi.infinite_of_left {ι : Type*} {π : ι → Sort*} [∀ i, nontrivial $ π i]
+  [infinite ι] : infinite (Π i : ι, π i) :=
+begin
+  choose m n hm using (λ i, exists_pair_ne (π i)),
```
```suggestion
  choose m n hm using λ i, exists_pair_ne (π i),
```

alreadydone commented on src/data/dfinsupp/basic.lean:
```diff
@@ -1929,3 +1935,31 @@ add_monoid_hom.congr_fun (comp_lift_add_hom h.to_add_monoid_hom g) f
+  fintype (Π₀ i, π i) :=
+fintype.of_equiv (Π i, π i) dfinsupp.equiv_fun_on_fintype.symm
+
+/-- Weaker form of `dfinsupp.fintype`. -/
```
I think it's not exactly "weaker", just that you get weaker conclusion from weaker assumptions.

urkud commented on src/data/dfinsupp/basic.lean:
```diff
@@ -1929,3 +1935,31 @@ add_monoid_hom.congr_fun (comp_lift_add_hom h.to_add_monoid_hom g) f
+  [finite ι] [∀ i, finite (π i)] :
+  finite (Π₀ i, π i) :=
+by letI := fintype.of_finite ι;
+exact finite.of_equiv (Π i, π i) dfinsupp.equiv_fun_on_fintype.symm
```
You can prove this for any `fun_like`.

Vierkantor committed:
Generalize `dfinsupp.finite` to all `fun_like`s

Vierkantor committed:
Unnecessary parens

Co-authored-by: Junyan Xu <junyanxumath@gmail.com>

alreadydone commented on src/data/fun_like/fintype.lean:
```diff
@@ -0,0 +1,62 @@
+ * `fun_like.fintype` is a **definition** stating all `fun_like`s are finite if their domain and
+   codomain are.
+   This is not an instance because specific `fun_like` types might have a better-suited definition.
+ * `fin_like.finite` is an **instance** stating all `fun_like`s are finite if their domain and
```
```suggestion
 * `fun_like.finite` is an **instance** stating all `fun_like`s are finite if their domain and
```

Vierkantor committed:
Typo fix

Co-authored-by: Junyan Xu <junyanxumath@gmail.com>

Vierkantor committed:
Add `dfinsupp.infinite_of_exists_right`

Vierkantor committed:
These can't be instances because they cause loops

alreadydone reviewed (approved):
Thanks!

alreadydone commented on src/data/fintype/basic.lean:
```diff
@@ -2130,6 +2130,36 @@ begin
+/-- See `pi.infinite_of_exists_right` for the case that only one `π i` is infinite. -/
+instance pi.infinite_of_right {ι : Type*} {π : ι → Sort*} [∀ i, infinite $ π i] [nonempty ι] :
+  infinite (Π i : ι, π i) :=
+pi.infinite_of_exists_right (classical.arbitrary ι)
```
I think we can just use `π : ι → Type*` everywhere since if for Prop the pi type is always a subsingleton. But I guess `ι : Sort*` could be useful? Same for `infinite` instances for `dfinsupp`.
```suggestion
instance pi.infinite_of_left {ι : Type*} {π : ι → Type*} [∀ i, nontrivial $ π i]
  [infinite ι] : infinite (Π i : ι, π i) :=
begin
  choose m n hm using λ i, exists_pair_ne (π i),
  refine infinite.of_injective (λ i, m.update i (n i)) (λ x y h, not_not.1 $ λ hne, _),
  simp_rw [update_eq_iff, update_noteq hne] at h,
  exact (hm x h.1.symm).elim,
end

/-- If at least one `π i` is infinite and the rest nonempty, the pi type of all `π` is infinite. -/
lemma pi.infinite_of_exists_right {ι : Type*} {π : ι → Type*} (i : ι)
  [infinite $ π i] [∀ i, nonempty $ π i] :
  infinite (Π i : ι, π i) :=
let ⟨m⟩ := @pi.nonempty ι π _ in infinite.of_injective _ (update_injective m i)

/-- See `pi.infinite_of_exists_right` for the case that only one `π i` is infinite. -/
instance pi.infinite_of_right {ι : Type*} {π : ι → Type*} [∀ i, infinite $ π i] [nonempty ι] :
  infinite (Π i : ι, π i) :=
pi.infinite_of_exists_right (classical.arbitrary ι)
```

Vierkantor replied:
I must say I'd much rather go the opposite way, generalizing rather than restricting the domain. In particular, I've had cases where I had to use `psigma` which takes values in a `Sort u` that is not defeq to a `Type v`, fighting tons of definitions unnecessarily restrict to `Type`s when they work just as well for `Sort`s.

alreadydone replied:
Oh yeah, this instance can probably apply to `Sort (u + 1)` (which is defeq to `Type u`) without problem, but would have trouble to apply to `Sort (max 1 u)`, so we should probably generalize.

alreadydone commented on src/data/fun_like/fintype.lean:
```diff
@@ -0,0 +1,62 @@
+
+Can't be an instance because it can cause infinite loops.
+-/
+lemma fun_like.finite [finite α] [Π i, finite (β i)] : finite F :=
```
Since it's a Prop:
```suggestion
lemma fun_like.finite [finite α] [∀ i, finite (β i)] : finite F :=
```

alreadydone commented on src/data/finsupp/fintype.lean:
```diff
@@ -0,0 +1,29 @@
+-/
+
+noncomputable instance finsupp.fintype {ι π : Sort*} [decidable_eq ι] [has_zero π]
+  [fintype ι] [fintype π]  :
```
```suggestion
  [fintype ι] [fintype π] :
```

alreadydone commented on src/data/fun_like/fintype.lean:
```diff
@@ -0,0 +1,62 @@
+They can't be instances themselves since they can cause loops.
+-/
+
+variables (F G : Type*) {α γ : Type*} {β : α → Type*} [fun_like F α β] [fun_like G α (λ _, γ)]
```
```suggestion
variables (F G α γ : Sort*} {β : α → Sort*} [fun_like F α β] [fun_like G α (λ _, γ)]
```

Vierkantor replied:
I would prefer to keep `α β γ` implicit here since they are determined by `F` and `G`:
```suggestion
variables (F G : Sort*) {α γ : Sort*} {β : α → Sort*} [fun_like F α β] [fun_like G α (λ _, γ)]
```

Vierkantor replied:
This one doesn't work because `fintype` is defined on `Type`, which in retrospect should have been obvious to me :)

alreadydone replied:
Oh sorry, I failed to notice the distinction betwen `()` and `{}` again. And yeah, `Sort*` doesn't work for `fintype`, but it works for `finite`.

Vierkantor committed:
Style fixes

Co-authored-by: Junyan Xu <junyanxumath@gmail.com>

Vierkantor committed:
Generalize to `Sort`

Vierkantor committed:
Revert "Generalize to `Sort`"

This reverts commit bd801cb4a77ab0dbb83d99b574295eaa6dc85b75.

Vierkantor committed:
Replace `Type` with `Sort` where appropriate

riccardobrasca commented on src/data/dfinsupp/basic.lean:
```diff
@@ -529,6 +529,12 @@ begin
 end
 
+/-- `dfinsupp.single a b` is injective in `a`. For the statement that it is injective in `b`, see
+`dfinsupp.single_injective` -/
```
Should `dfinsupp.single_injective` be called `dfinsupp.single_right_injective` now?

Vierkantor replied:
`dfinsupp.single_injective` matches `finsupp.single_injective` and `pi.single_injective` and all of these occur in a lot of places. I'm not opposed to changing this, but I'd rather do it in a follow-up PR.

riccardobrasca reviewed (approved):
Thanks!

bors d+
